quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/filter/KeycloakGracefulShutdownFilter.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright 2025 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.quarkus.runtime.filter;
+
+import io.netty.handler.codec.http.HttpHeaderNames;
+import io.quarkus.runtime.shutdown.ShutdownListener;
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+import io.vertx.core.http.HttpVersion;
+import io.vertx.core.http.impl.Http1xServerConnection;
+import io.vertx.core.http.impl.Http2ServerConnection;
+import io.vertx.ext.web.RoutingContext;
+import jakarta.enterprise.inject.spi.CDI;
+import org.jboss.logging.Logger;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Provide a graceful shutdown. Inspired by {@link io.quarkus.vertx.http.runtime.filters.GracefulShutdownFilter}.
+ * <p>
+ * This explores options while we wait for <a href="https://github.com/keycloak/keycloak/issues/43589">#43589</a>.
+ */
+public class KeycloakGracefulShutdownFilter implements ShutdownListener, Handler<RoutingContext> {
+
+    private static final Logger log = Logger.getLogger(KeycloakGracefulShutdownFilter.class);
+
+    private volatile boolean running = true;
+    private final AtomicInteger currentRequestCount = new AtomicInteger();
+    private final AtomicReference<ShutdownNotification> notification = new AtomicReference<>();
+
+    private final Handler<AsyncResult<Void>> requestDoneHandler = event -> {
+        int count = currentRequestCount.decrementAndGet();
+        if (!running) {
+            if (count == 0) {
+                executeNotification();
+            }
+        }
+    };
+    private Long timer;
+
+    private void executeNotification() {
+        ShutdownNotification n = notification.get();
+        if (n != null) {
+            if (notification.compareAndSet(n, null)) {
+                n.done();
+                log.info("All HTTP requests complete");
+                if (timer != null) {
```

Comment:
Is timer cancellation required? If so, there's a possible race-condition where cancellation is skipped.

Reply:
No, cancellation is not required in vertx